src/server/commands/mod.rs

```diff
@@ -1,9 +1,103 @@
+use crate::server::error::FTPError;
 use crate::server::password::Password;
+use crate::server::reply::Reply;
+use crate::server::CommandArgs;
+use crate::storage;
 
 use bytes::Bytes;
 use failure::*;
 use std::{fmt, result, str};
 
+#[macro_use]
+mod macros;
+
+mod abor;
+mod acct;
+mod allo;
+mod auth;
+mod ccc;
+mod cdc;
+mod cdup;
+mod cwd;
+mod dele;
+mod feat;
+mod help;
+mod list;
+mod mdtm;
+mod mkd;
+mod mode;
+mod nlst;
+mod noop;
+mod opts;
+mod pass;
+mod pasv;
+mod pbsz;
+mod port;
+mod prot;
+mod pwd;
+mod quit;
+mod rest;
+mod retr;
+mod rmd;
+mod rnfr;
+mod rnto;
+mod size;
+mod stat;
+mod stor;
+mod stou;
+mod stru;
+mod syst;
+mod type_;
+mod user;
+
+pub use abor::Abor;
+pub use acct::Acct;
+pub use allo::Allo;
+pub use auth::Auth;
+pub use ccc::Ccc;
+pub use cdc::Cdc;
+pub use cdup::Cdup;
+pub use cwd::Cwd;
+pub use dele::Dele;
+pub use feat::Feat;
+pub use help::Help;
+pub use list::List;
+pub use mdtm::Mdtm;
+pub use mkd::Mkd;
+pub use mode::Mode;
+pub use nlst::Nlst;
+pub use noop::Noop;
+pub use opts::Opts;
+pub use pass::Pass;
+pub use pasv::Pasv;
+pub use pbsz::Pbsz;
+pub use port::Port;
+pub use prot::Prot;
+pub use pwd::Pwd;
+pub use quit::Quit;
+pub use rest::Rest;
+pub use retr::Retr;
+pub use rmd::Rmd;
+pub use rnfr::Rnfr;
+pub use rnto::Rnto;
+pub use size::Size;
+pub use stat::Stat;
+pub use stor::Stor;
+pub use stou::Stou;
+pub use stru::Stru;
+pub use syst::Syst;
+pub use type_::Type;
+pub use user::User;
+
+pub trait Cmd<S, U: Send + Sync>
```

Comment:
Since I'm employing the _Command_ pattern, and this trait defines the contract of a _Command_, I do not agree with this suggestion. A command handler would be some piece of code that executes the command (like `Server` is, now). Rather, I think we should work to remove the `Command` enum instead.

Reply:
Remember FTP has the concept of a `Command` see [the RFC](https://tools.ietf.org/html/rfc959). So by elevating the importance of the `Command` pattern you're obscuring the domain specific concept of a Command.  

See Uncle Bob's [writeup of the Screaming Architecture](https://blog.cleancoder.com/uncle-bob/2011/09/30/Screaming-Architecture.html) so see where I'm coming from. 

An FTP server architecture should scream domain specific things like command, reply etc rather than GOF patterns.

So when I mentioned this I saw for instance a MKD command (domain specific concept) and the piece of code that turns the command sent from the client into a response as the `Command Handler` regardless of the implementation pattern it employs.  I also had in my head the Go handler:

```
type Handler interface {
	ServeHTTP(ResponseWriter, *Request)
}
```

that takes an already parsed HTTP request and turns it into a response.

Be that as it may, this can be changed later too, or discussed and not changed. Lets hear what others think.

Reply:
To me, the GoF Command pattern does not shadow the concept of the FTP command. In fact, I think they are one and the same thing. It is a domain specific concept of a command, implemented as a GoF Command. It's not a coincidence that these both have the same name.

Reply:
Differences aside, this code is already 1000 x better than what we had so I'm happy if we merge.

Reply:
Differences aside, this code is already 1000 x better than what we had so I'm happy if we merge.